### PR TITLE
Remove use of `ExceptT`

### PR DESCRIPTION
**Pull request by edsko**
This big refactoring allows the mock file system and the immutable DB to be run on bare `IO`, using exceptions for error reporting, whilst at the same time allowing the pure models to use `ExceptT` where needed. 

This should make @avieth happy :)

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/API.hs [hunk 1]
```diff
@@ -151,7 +148,10 @@ data ImmutableDB m = ImmutableDB
       :: HasCallStack
       => EpochSlot
       -> EpochSlot
-      -> m (Either ImmutableDBError (Iterator m))
+      -> m (Iterator m)
+
+    -- Throw immutable DB errors
+  , immutableDbErr :: ErrorHandling ImmutableDBError m
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/API.hs]: Suggestion:
```haskell
    -- | How to throw an 'ImmutableDBError'.
  , hasImmutableDBErr :: ErrorHandling ImmutableDBError m
```

Also, I have noticed we don't actually use this right now. But I guess I might need it in my q-s-m tests.

File: ouroboros-consensus/src/Ouroboros/Storage/FS/IO.hs [hunk 2]
```diff
@@ -1,22 +1,16 @@
 {-# LANGUAGE BangPatterns      #-}
 {-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RecordWildCards   #-}
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/FS/IO.hs]: HLint tells me you don't need `-XRecordWildCards` here.

File: ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Types.hs [hunk 3]
```diff
@@ -68,6 +69,8 @@ data ImmutableDBError
     -- layer.
   deriving (Show, Generic)
 
+instance Exception ImmutableDBError
+
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Types.hs]: Maybe use `prettyImmutableDBError` instead of `show`:
```haskell
    displayException = prettyImmutableDBError
```

File: ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Util.hs [hunk 4]
```diff
@@ -31,18 +33,36 @@ import           Ouroboros.Storage.ImmutableDB.Types
...
-throwUserError :: (HasCallStack, Monad m)
-               => UserError -> ExceptT ImmutableDBError m a
-throwUserError ue = throwError $ UserError ue (popCallStack callStack)
--- 'popCallStack' to remove the call to 'throwError' from the call stack
-
-throwUnexpectedError :: Monad m
-                     => UnexpectedError -> ExceptT ImmutableDBError m a
-throwUnexpectedError ue = throwError $ UnexpectedError ue
-
+handleFsError :: ErrorHandling ImmutableDBError m
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Util.hs]: This function is not used at the moment.

File: ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Util.hs [hunk 5]
```diff
@@ -31,18 +33,36 @@ import           Ouroboros.Storage.ImmutableDB.Types
...
+handleUser :: HasCallStack
+           => ErrorHandling ImmutableDBError m
+           -> ErrorHandling UserError        m
+handleUser = EH.embed (flip UserError (popCallStack callStack)) $ \case
+               UserError e _ -> Just e
+               _otherwise    -> Nothing
+
+handleUnexpected :: ErrorHandling ImmutableDBError m
+                 -> ErrorHandling UnexpectedError  m
+handleUnexpected = EH.embed (UnexpectedError) $ \case
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Util.hs]: You can leave out the parentheses around `(UnexpectedError)`.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/Util.hs [hunk 6]
```diff
@@ -83,80 +88,106 @@ expectUnexpectedError unexpectetdErrPred = expectDBError $ \case
...
+expectResult :: (e -> String)
+             -> (a -> Assertion)
              -> Either e a
              -> Assertion
-expectResult prettyError expectedResult =
+expectResult prettyError checkResult =
     \case
        Left e -> fail ("Unexpected error: " <> prettyError e)
-       Right actualResult -> actualResult @?= expectedResult
+       Right actualResult -> checkResult actualResult -- @?= expectedResult
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/Util.hs]: What's the reason for this change, I assume a missing `Eq a` or `Show a`?

You can the delete the commented-out code.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/Util.hs [hunk 7]
```diff
@@ -83,80 +88,106 @@ expectUnexpectedError unexpectetdErrPred = expectDBError $ \case
...
+
+tryFS :: IO a -> IO (Either FsError a)
+tryFS = E.try
+
+tryImmDB :: IO a -> IO (Either ImmutableDBError a)
+tryImmDB = fmap squash . E.try . E.try
+  where
+    -- TODO: With the redesigned error handling I'm not sure whether it's
+    -- still necessary that ImmutableDBError can wrap FsError; I think we can
+    -- get rid of that. If we do, this function should become
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/Util.hs]: But what if you want to do it in a *pure* way, e.g., `Either ImmutableDBError a`? You'd still need a way to embed the `FsError` in `ImmutableDBError`, right? However, maybe we never even need to run it in a *pure* way, as we use `IO` for `MonadSTM` anyway.

File: ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Util.hs (hunk 4 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Util.hs]: Hmmm. Ok, will delete it. Odd :thinking:

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/Util.hs (hunk 6 above)
Comment (edsko) [reply ouroboros-consensus/test-storage/Test/Ouroboros/Storage/Util.hs]: No. The reason is that a side effect of this reshuffling is that we now get `Either FsError (SomeResult, MockFS` as the result from running the mock file system, as opposed to `(Either FsError SomeResult, MockFs)`. This doesn't really matter, many of the tests only want to assert something about that `SomeResult` rather than the final `MockFS`, and so they now look like `(@?= ..) . fst`.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/Util.hs (hunk 7 above)
Comment (edsko) [reply ouroboros-consensus/test-storage/Test/Ouroboros/Storage/Util.hs]: No, that's not needed. You can just run `ExceptT (Either FsError ImmutableError)` and that would be perfectly fine.

**Comment by edsko** [comment]
@mrBliss addressed your feedback and squashed all commits.
